README.md
# Interactive Add button layout

Custom Layout with interactive add button to impove your UI and UX .

the package is available [here](https://pub.dev/packages/interactive_add_button_layout)

inspired from [Oleg Frolov](https://dribbble.com/shots/6558740-Add-Button-Interaction).

## Usage 

### Import the Package 
add this dependencies to your app
` dependencies: interactive_add_button_layout: ^0.1.0 `
### Use the Package
add this import statement 
` import 'package:interactive_add_button_layout/interactive_add_button_layout.dart';`

The layout need to be the root layout of your widget (screen)

and Now to use it, add this code to your widget : 

```Dart
return Scaffold(
      ...
      body: AddButtonLayout( parameters )

```
The layout has 6 parameters which are : 

* `child` : you know what is that xD, in case you don't it's the child of the layout which mean that the layout is his parent .
* `row` : a List of Widgets to be diplayed in a row for the Row layout .
* `column` : a List of Widgets to be diplayed in a column for the Column layout .
* `onPressed`: the function to be called when the user click the add button .
* `color` : the color of the layout (color of the background), by default it's `Color(0xff2A1546)` .
* `btnColor` : the color of the add button 

the `row` and `column` and `child` are required !

## Example : 
you can find a demo app in `./example`

## Gif 

<img src="https://github.com/DokkarRachidReda/Interactive-Add-Button-Layout/blob/master/add_button.gif" width="300"/>

## Contribution 
Feel free to contribute, to report a bug or  to suggest a feature,Thank you :)


CHANGELOG.md
## [0.1.0] : 24/8/2019.

* First  release.


test/interactive_add_button_layout_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:interactive_add_button_layout/interactive_add_button_layout.dart';


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:interactive_add_button_layout/interactive_add_button_layout.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'interactive add button layout Demo',
      home: MyHomePage(title: 'interactive add button layout Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
        backgroundColor: Color(0xff0F051F),
        body: AddButtonLayout(
          child: Padding(
            padding: EdgeInsets.only(top: 80, left: 10),
            child: Column(
              children: <Widget>[
                myWidget(),
                myWidget(),
                myWidget(),
                myWidget(),
              ],
            ),
          ),
          row: <Widget>[
            Padding(
              padding: EdgeInsets.only(top: 150),
            ),
            Container(
              width: 70,
              height: 60,
              padding: EdgeInsets.all(15),
              decoration: BoxDecoration(
                  borderRadius: BorderRadius.circular(10),
                  color: Color(0xfffdfdfd).withOpacity(0.7)),
              child: Image(
                image: AssetImage("./assets/camera.png"),
              ),
            ),
            Padding(
                padding: EdgeInsets.only(top: 10, bottom: 10),
                child: Container(
                  width: 70,
                  height: 60,
                  padding: EdgeInsets.all(15),
                  decoration: BoxDecoration(
                      borderRadius: BorderRadius.circular(10),
                      color: Color(0xfffdfdfd).withOpacity(0.7)),
                  child: Image(
                    image: AssetImage("./assets/camera.png"),
                  ),
                )),
            Container(
              width: 70,
              height: 60,
              padding: EdgeInsets.all(15),
              decoration: BoxDecoration(
                  borderRadius: BorderRadius.circular(10),
                  color: Color(0xfffdfdfd).withOpacity(0.7)),
              child: Image(
                image: AssetImage("./assets/camera.png"),
              ),
            )
          ],
          column: <Widget>[
            Padding(
              padding: EdgeInsets.only(bottom: 5),
              child: Container(
                height: 45,
                width: 250,
                decoration: BoxDecoration(
                    borderRadius: BorderRadius.circular(60),
                    color: Color(0xfffdfdfd).withOpacity(0.7)),
              ),
            )
          ],
        ));
  }

  Widget myWidget() {
    return Container(
        height: 120,
        width: 350,
        padding: EdgeInsets.all(5),
        child: Padding(
          padding: EdgeInsets.fromLTRB(2, 2, 0, 0),
          child: Row(
            children: <Widget>[
              Padding(
                padding: EdgeInsets.only(bottom: 30, left: 10),
                child: Container(
                  width: 50,
                  height: 50,
                  decoration: BoxDecoration(
                      color: Color(0xffEBC0D2).withOpacity(0.5),
                      shape: BoxShape.circle),
                ),
              ),
              Padding(
                  padding: EdgeInsets.only(left: 10, top: 10),
                  child: Container(
                    height: 150,
                    width: 250,
                    decoration: BoxDecoration(
                        borderRadius: BorderRadius.circular(20),
                        color: Color(0xffEBC0D2).withOpacity(0.2)),
                    child: Column(
                      children: <Widget>[
                        Padding(
                          padding: EdgeInsets.only(top: 10, right: 10),
                          child: Container(
                            width: 200,
                            height: 20,
                            decoration: BoxDecoration(
                              borderRadius: BorderRadius.circular(60),
                              color: Color(0xfdedede).withOpacity(0.4),
                            ),
                          ),
                        ),
                        Padding(
                          padding: EdgeInsets.only(
                            top: 10,
                          ),
                          child: Container(
                            width: 220,
                            height: 20,
                            decoration: BoxDecoration(
                              borderRadius: BorderRadius.circular(60),
                              color: Color(0xfdedede).withOpacity(0.3),
                            ),
                          ),
                        ),
                        Padding(
                          padding: EdgeInsets.only(
                            top: 10,
                          ),
                          child: Container(
                            width: 220,
                            height: 20,
                            decoration: BoxDecoration(
                              borderRadius: BorderRadius.circular(60),
                              color: Color(0xfdedede).withOpacity(0.3),
                            ),
                          ),
                        ),
                      ],
                    ),
                  ))
            ],
          ),
        ));
  }
}


lib/interactive_add_button_layout.dart
library interactive_add_button_layout;

import 'package:flutter/material.dart';

class AddButtonLayout extends StatefulWidget {
  const AddButtonLayout(
      {@required this.child,
      @required this.row,
      @required this.column,
      this.color = const Color(0xff2A1546),
      this.btnColor = const Color(0xffb31634),
      this.onPressed});

  final Widget child;
  final List<Widget> row;
  final List<Widget> column;
  final Color color;
  final Color btnColor;
  final Function onPressed;
  @override
  AdlState createState() {
    return AdlState();
  }
}

class AdlState extends State<AddButtonLayout>
    with SingleTickerProviderStateMixin {
  AnimationController ac;
  Animation animation;
  bool isOpnen = false;
  @override
  void initState() {
    ac =
        AnimationController(vsync: this, duration: Duration(milliseconds: 500));
    ac.addListener(() {
      setState(() {});
    });
    animation = Tween<double>(begin: 0.0, end: -0.3).animate(CurvedAnimation(
      curve: Curves.easeInOutSine,
      parent: ac,
    ));
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    double width = MediaQuery.of(context).size.width;
    return Container(
      child: Stack(
        children: <Widget>[
          Container(
            alignment: Alignment.bottomRight,
            child: Column(
              mainAxisAlignment: MainAxisAlignment.center,
              children: widget.row,
            ),
          ),
          Container(
            alignment: Alignment.bottomCenter,
            child: Row(
              mainAxisAlignment: MainAxisAlignment.center,
              children: widget.column,
            ),
          ),
          Container(
            width: width,
            height: MediaQuery.of(context).size.height,
            color: widget.color,
            child: widget.child,
            transform: Matrix4.translationValues(
                width * animation.value, width * animation.value * 0.7, 0),
          ),
          Container(
              alignment: Alignment.bottomRight,
              child: Padding(
                padding: EdgeInsets.only(right: 15, bottom: 10),
                child: FloatingActionButton(
                    backgroundColor: widget.btnColor,
                    onPressed: () {
                      !isOpnen ? ac.forward() : ac.reverse();
                      isOpnen ? isOpnen = false : isOpnen = true;
                      widget.onPressed != null
                          ? widget.onPressed()
                          : print("on Pressed is null");
                    },
                    child: RotationTransition(
                      turns: Tween(begin: 0.0, end: 0.4).animate(ac),
                      child: Icon(Icons.add),
                    )),
              ))
        ],
      ),
    );
  }
}
